File: server/src/protocol/CommandFactory.cpp
#include "CommandFactory.h"
#include "BlindsDOWNOnTimeBuilder.h"
#include "BlindsUPBuilder.h"
#include "BlindsStatusBuilder.h"
#include "BlindsDOWNBuilder.h"
#include "BlindsUPOnTimeBuilder.h"
#include "AutomaticBlindsBuilder.h"
#include "TemperatureDemandBuilder.h"
#include "UserOutOfHomeBuilder.h"
#include "UserInHomeBuilder.h"
#include "EndConnectionBuilder.h"
#include "SnapshotBuilder.h"
#include "StartStreamBuilder.h"
#include "StopStreamBuilder.h"
#include "GuardStatusCommandBuilder.h"
#include "MotorStatusCommandBuilder.h"
#include "SensorTagSamplesBuilder.h"

#include <stdexcept>

using namespace std;
using namespace communication;

CommandFactory::CommandFactory()
{}

CommandFactory::~CommandFactory()
{}

unique_ptr<Command> CommandFactory::createCommand(const vector<uint8_t> &commandInBytes)
{
    const auto type = static_cast<CommandType>(commandInBytes[COMMAND_TYPE_POSITION]);

    switch (type)
    {
        case CommandType::BLINDS_UP :
            builder_ = make_unique<BlindsUPBuilder>();
            return move(builder_->create(commandInBytes));
        case CommandType::BLINDS_DOWN :
            builder_ = make_unique<BlindsDOWNBuilder>();
            return move(builder_->create(commandInBytes));
        case CommandType::BLINDS_UP_ON_TIME :
            builder_ = make_unique<BlindsUPOnTimeBuilder>();
            return move(builder_->create(commandInBytes));
        case CommandType::BLINDS_DOWN_ON_TIME :
            builder_ = make_unique<BlindsDOWNOnTimeBuilder>();
            return move(builder_->create(commandInBytes));
        case CommandType::BLINDS_STATUS :
            builder_ = make_unique<BlindsStatusBuilder>();
            return move(builder_->create(commandInBytes));
        case CommandType::AUTOMATIC_BLINDS :
            builder_ = make_unique<AutomaticBlindsBuilder>();
            return move(builder_->create(commandInBytes));
        case CommandType::TEMPERATURE_DEMAND :
            builder_ = make_unique<TemperatureDemandBuilder>();
            return move(builder_->create(commandInBytes));
        case CommandType::USER_OUT_OF_HOME :
            builder_ = make_unique<UserOutOfHomeBuilder>();
            return move(builder_->create(commandInBytes));
		case CommandType::USER_IN_HOME :
            builder_ = make_unique<UserInHomeBuilder>();
            return move(builder_->create(commandInBytes));
        case CommandType::SNAPSHOT :
            builder_ = make_unique<SnapshotBuilder>();
            return move(builder_->create(commandInBytes));
        case CommandType::START_STREAM :
            builder_ = make_unique<StartStreamBuilder>();
            return move(builder_->create(commandInBytes));
        case CommandType::STOP_STREAM :
            builder_ = make_unique<StopStreamBuilder>();
            return move(builder_->create(commandInBytes));
		case CommandType::MOTOR_STATUS_COMMAND :
            builder_ = make_unique<MotorStatusCommandBuilder>();
            return move(builder_->create(commandInBytes));
		case CommandType::GUARD_STATUS_COMMAND :
            builder_ = make_unique<GuardStatusCommandBuilder>();
            return move(builder_->create(commandInBytes));
        case CommandType::END_CONNECTION :
            builder_ = make_unique<EndConnectionBuilder>();
            return move(builder_->create(commandInBytes));
        case CommandType::SENSOR_TAG_SAMPLES :
            builder_ = make_unique<SensorTagSamplesBuilder>();
            return move(builder_->create(commandInBytes));
        default:
            throw invalid_argument("Received command does not register in factory.");

    }
}

File: mobile_client/app/src/main/java/wpam/mobile_client/protocol/CommandFactory.java
package wpam.mobile_client.protocol;
import java.io.*;
import java.util.ArrayList;

public class CommandFactory implements Closeable
	{
		private CommandBuilder builder;
		private final int COMMAND_TYPE_POSITION = 5;

		public CommandFactory()
		{
		}
		public final void close()
		{
		}

		public Command createCommand(ArrayList<Integer> commandInBytes)
		{
			CommandType type = CommandType.forValue(commandInBytes.get(COMMAND_TYPE_POSITION));
			Command command;
			switch (type)
			{
				case BLINDS_DOWN:
				{
					this.builder = new BlindsDOWNBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case BLINDS_DOWN_ON_TIME:
				{
					this.builder = new BlindsDOWNOnTimeBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case BLINDS_UP:
				{
					this.builder = new BlindsUPBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case BLINDS_UP_ON_TIME:
				{
					this.builder = new BlindsUPOnTimeBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case SNAPSHOT:
				{
					this.builder = new SnapshotBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case AUTOMATIC_BLINDS:
				{
					this.builder = new AutomaticBlindsBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case BLINDS_STATUS:
				{
					this.builder = new BlindsStatusBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case USER_OUT_OF_HOME:
				{
					this.builder = new UserOutOfHomeBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case USER_IN_HOME:
				{
					this.builder = new UserInHomeBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case TEMPERATURE_DEMAND:
				{
					this.builder = new TemperatureDemandBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case END_CONNECTION:
				{
					this.builder = new EndConnectionBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case START_STREAM:
				{
					this.builder = new StartStreamBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case STOP_STREAM:
				{
					this.builder = new StopStreamBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case GUARD_STATUS_COMMAND:
				{
					this.builder = new GuardStatusCommandBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case MOTOR_STATUS_COMMAND:
				{
					this.builder = new MotorStatusCommandBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case SENSOR_TAG_SAMPLES:
				{
					this.builder = new SensorTagSamplesBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case PLOT:
				{
					this.builder = new PlotBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				default:
				{
					throw new IllegalArgumentException("Score is not valid!");
				}
			}
			return command;
		}
	}